nexus/src/db/datastore.rs

```diff
@@ -2833,6 +2834,42 @@ impl DataStore {
 
         Ok(())
     }
+
+    /// Return the next available IPv6 address for an Oxide service running on
+    /// the provided sled.
+    pub async fn next_ipv6_address(
+        &self,
+        opctx: &OpContext,
+        sled_id: Uuid,
+    ) -> Result<Ipv6Addr, Error> {
+        use db::schema::sled::dsl;
+        let net = diesel::update(
+            dsl::sled.find(sled_id).filter(dsl::time_deleted.is_null()),
+        )
+        .set(dsl::last_used_address.eq(dsl::last_used_address + 1))
+        .returning(dsl::last_used_address)
+        .get_result_async(self.pool_authorized(opctx).await?)
+        .await
+        .map_err(|e| {
+            public_error_from_diesel_pool(
+                e,
+                ErrorHandler::NotFoundByLookup(
+                    ResourceType::Sled,
+                    LookupType::ById(sled_id),
+                ),
+            )
+        })?;
+
+        // TODO-correctness: We need to ensure that this address is actually
+        // within the sled's underlay prefix, once that's included in the
+        // database record.
```

Comment:
It does seem easy and valuable to check that we're not bumping the addresses past the sled's `/64`. I think I avoided doing that now because we don't yet have the underlay prefix in the database record, and it's also possible that the sled server could be run with an IPv4 address. I've got some changes in the works to move everything to IPv6 only, but until that happens, I wasn't sure what check to put if the address happens to be IPv4. I was just hoping to do that switch to V6 everywhere concurrently with adding the sled's underlay prefix, to keep this PR small.